src/plugins/embeddable/public/lib/state_transfer/embeddable_state_transfer.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { cloneDeep } from 'lodash';
+import { CoreStart, ScopedHistory } from '../../../../../core/public';
+import {
+  EmbeddableOriginatingAppState,
+  isEmbeddableOriginatingAppState,
+  EmbeddablePackageState,
+  isEmbeddablePackageState,
+} from './types';
+
+export class EmbeddableStateTransfer {
+  constructor(private coreStart: CoreStart) {}
+
+  public incomingOriginatingApp(
+    history: ScopedHistory,
+    removeAfterFetch: boolean = false
+  ): EmbeddableOriginatingAppState | undefined {
+    return this.incoming<EmbeddableOriginatingAppState>(
+      history,
+      isEmbeddableOriginatingAppState,
+      removeAfterFetch
+    );
+  }
+
+  public incomingEmbeddablePackage(
+    history: ScopedHistory,
+    removeAfterFetch: boolean = false
+  ): EmbeddablePackageState | undefined {
+    return this.incoming<EmbeddablePackageState>(
+      history,
+      isEmbeddablePackageState,
+      removeAfterFetch
+    );
+  }
+
+  public incoming<IncomingStateType>(
```

Comment:
Originally I was thinking of leaving these public for use with custom types, but you make a good point above about this being specific to embeddable.

Those generic methods being public might make sense in a different implementation where:
- the state transfer service consists of just those two methods (renamed getTypedState & navigateToWithTypedState maybe), and lives in the `core.application` service
- the embeddable plugin contains the types `EmbeddablePackageState`, and `OriginatingAppState`
- The plugins that use this API do something like 
``` tsx
import { ......, EmbeddablePackageState } from '../../../embeddable/public';

...
// outgoing
application.getTypedState<EmbeddablePackageState>(this.scopedHistory());

```
and:

``` tsx
import { ......, EmbeddablePackageState } from '../../../embeddable/public';

...
// incoming
const state = { type: TYPE, id };
application.navigateToWithTypedState<EmbeddablePackageState>(
  this.originatingApp, 
  { state }
);

```

This could be a quick change, and would make this solution more generic but would add to the core API. Do you think this is a change worth making?

For now though, I've made the two generic methods private.

Reply:
> This could be a quick change, and would make this solution more generic but would add to the core API. Do you think this is a change worth making?

I wouldn't do that in this PR.  I have concerns about the types giving consumers false safety.  A consumer could call:

```
const state = application.getTypedState<EmbeddablePackageState>(this.scopedHistory());
```

But how do they know that scoped history is typed that way?  Because `this.scopedHistory()` is only known at run time, that isn't type safe.

I could be wrong but it'd be easier to discuss in a dedicated PR. So, ++ for keeping the generic typed functions private in this PR.